ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -545,6 +556,16 @@ private StreamToTableJoiner(
                 contextStacker
             );
           }
+        case RIGHT:
+          if (joinKey.isForeignKey()) {
+            throw new KsqlException("RIGHT OUTER JOIN on a foreign key is not supported");
```

Comment:
I could not find a way to implement FK right join given the existing primitives. As the scope of the issue is to provide syntactical sugar in KSQL rather than to extend KS joins, I decided to throw an exception here.

The issue with supporting FK right join is that KS implementation of FK join is not really symmetric. Let's look at the foreign key left join:

https://github.com/confluentinc/ksql/blob/a2efcc53f05de45daf0724ff7c4e1467f2548224/ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/ForeignKeyTableTableJoinBuilder.java#L110-L114

Several observations:
* The types are not symmetric. Table-Table join has a type `TableTableJoin<K>` while Table-Table foreign key join has a type `ForeignKeyTableTableJoin<KLeftT, KRightT>`.
* KS extracts the FK from the RHS of the expression. It is not possible to extract FK from the primary table.
* On the KS side, the join behavior is controlled by `SubscriptionWrapper#Instruction`:

```java
    public enum Instruction {
        //Send nothing. Do not propagate.
        DELETE_KEY_NO_PROPAGATE((byte) 0x00),

        //Send (k, null)
        DELETE_KEY_AND_PROPAGATE((byte) 0x01),

        //(changing foreign key, but FK+Val may not exist)
        //Send (k, fk-val) OR
        //Send (k, null) if fk-val does not exist
        PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE((byte) 0x02),

        //(first time ever sending key)
        //Send (k, fk-val) only if fk-val exists.
        PROPAGATE_ONLY_IF_FK_VAL_AVAILABLE((byte) 0x03);
```

Given that, I do not see a way to implement the right FK join with what we have at hand. I can work on a supporting right FK join, but it seems to dramatically extend the scope of the issue. Thinking about a proper fix, in my opinion, it should look as follows:

* Introduce a new instruction that sends `(k, fk-val)` or `(null, fk-val)`
* Extend `SubscriptionJoinForeignProcessorSupplier` to support the new instruction
* Introduce new KTable method(s) for right FK join. While at it we can also introduce a "usual" right join. This step will require a KIP.

Reply:
> As the scope of the issue is to provide syntactical sugar in KSQL rather than to extend KS joins, I decided to throw an exception here.

Well, adding `dimensionTable RIGHT JOIN factTable` is still syntactic sugar as it computes the same result as `factTable LEFT JOIN dimensionTable` -- so it is not extending joins at all.

In the end, we just need to "flip" left and right table, and convert `RIGHT JOIN` to `LEFT JOIN`. The rest should fall into place automatically?

If it too cumbersome given the current code base, we can still fix it in the logical planner rewrite I started, so no big deal.

Reply:
Is there an issue tracking support for right joins against foreign keys?